clash-protocols/src/Protocols/PacketStream/Base.hs

```diff
@@ -0,0 +1,554 @@
+
+{- |
+Strips trailing zero-byte transfers from packets in the stream. That is,
+if a packet consists of more than one transfer and '_last' of the last
+transfer in that packet is @Just 0@, the last transfer of that packet will
+be dropped and '_last' of the transfer before that will be set to @maxBound@.
+If such a trailing zero-byte transfer had '_abort' asserted, it will be
+preserved.
+
+Has one clock cycle latency, but runs at full throughput.
+-}
+stripTrailingEmptyC ::
+  forall (dataWidth :: Nat) (meta :: Type) (dom :: Domain).
+  (HiddenClockResetEnable dom) =>
+  (KnownNat dataWidth) =>
+  (NFDataX meta) =>
+  Circuit
+    (PacketStream dom dataWidth meta)
+    (PacketStream dom dataWidth meta)
+stripTrailingEmptyC = forceResetSanity |> fromSignals (mealyB go (False, False, Nothing))
+ where
+  go (notFirst, flush, cache) (Nothing, bwdIn) =
+    ((notFirst, flush', cache'), (PacketStreamS2M True, fwdOut))
+   where
+    fwdOut = if flush then cache else Nothing
+    (flush', cache')
+      | flush && _ready bwdIn = (False, Nothing)
+      | otherwise = (flush, cache)
+  go (notFirst, flush, cache) (Just transferIn, bwdIn) = (nextStOut, (bwdOut, fwdOut))
+   where
+    (notFirst', flush', cache', fwdOut) = case _last transferIn of
+      Nothing -> (True, False, Just transferIn, cache)
+      Just i ->
+        let trailing = i == 0 && notFirst
+         in ( False
+            , not trailing
+            , if trailing then Nothing else Just transferIn
+            , if trailing
+                then (\x -> x{_last = Just maxBound, _abort = _abort x || _abort transferIn}) <$> cache
+                else cache
+            )
+
+    bwdOut = PacketStreamS2M (Maybe.isNothing cache || _ready bwdIn)
+
+    nextStOut
+      | Maybe.isNothing cache || _ready bwdIn = (notFirst', flush', cache')
+      | otherwise = (notFirst, flush, cache)
+
+-- | Sets data bytes that are not enabled in a @PacketStream@ to @0x00@.
+zeroOutInvalidBytesC ::
+  forall (dom :: Domain) (dataWidth :: Nat) (meta :: Type).
+  (KnownNat dataWidth) =>
+  (1 <= dataWidth) =>
+  Circuit
+    (PacketStream dom dataWidth meta)
+    (PacketStream dom dataWidth meta)
+zeroOutInvalidBytesC = Circuit $ \(fwdIn, bwdIn) -> (bwdIn, fmap (go <$>) fwdIn)
+ where
+  go transferIn = transferIn{_data = dataOut}
+   where
+    dataOut = case _last transferIn of
+      Nothing -> _data transferIn
+      Just i ->
+        imap
+          (\(j :: Index dataWidth) byte -> if resize j < i then byte else 0x00)
+          (_data transferIn)
+
+{- |
+Copy data of a single `PacketStream` to multiple. LHS will only receive
+an acknowledgement when all RHS receivers have acknowledged data.
+-}
+fanout ::
+  forall n dataWidth meta dom.
+  (HiddenClockResetEnable dom) =>
+  (KnownNat n) =>
+  (KnownNat dataWidth) =>
+  (1 <= n) =>
+  (NFDataX meta) =>
+  Circuit
+    (PacketStream dom dataWidth meta)
+    (Vec n (PacketStream dom dataWidth meta))
+fanout = DfConv.fanout Proxy Proxy
```

Comment:
This doesn't preserve expected `PacketStream` semantics. It will distribute fragments of the same packet over multiple consumers. Something you most likely do not want to do. This should be explicit in the documentation. Persoanally I wouldn't export it all specialized for `PacketStream`